File: src/project_manager_flutter/lib/features/organizations/repository/organization_repository.dart
import 'package:dio/dio.dart';
import '../../../core/constants/api_constants.dart';
import '../../../core/network/api_exceptions.dart';
import '../models/create_organization_request.dart';
import '../models/organization.dart';
import '../models/organization_detail.dart';
import '../models/update_organization_request.dart';

class OrganizationRepository {
  final Dio _dio;

  OrganizationRepository(this._dio);

  Future<List<Organization>> listOrganizations() async {
    try {
      final response = await _dio.get(ApiConstants.organizations);
      return (response.data as List)
          .map((e) => Organization.fromJson(e))
          .toList();
    } on DioException catch (e) {
      throw ApiException.fromResponse(
        e.response?.statusCode ?? 500,
        e.response?.data,
      );
    }
  }

  Future<OrganizationDetail> getOrganization(String id) async {
    try {
      final response = await _dio.get(ApiConstants.organizationById(id));
      return OrganizationDetail.fromJson(response.data);
    } on DioException catch (e) {
      throw ApiException.fromResponse(
        e.response?.statusCode ?? 500,
        e.response?.data,
      );
    }
  }

  Future<Organization> createOrganization(CreateOrganizationRequest request) async {
    try {
      final response = await _dio.post(
        ApiConstants.organizations,
        data: request.toJson(),
      );
      return Organization.fromJson(response.data);
    } on DioException catch (e) {
      throw ApiException.fromResponse(
        e.response?.statusCode ?? 500,
        e.response?.data,
      );
    }
  }

  Future<Organization> updateOrganization(
      String id, UpdateOrganizationRequest request) async {
    try {
      final response = await _dio.put(
        ApiConstants.organizationById(id),
        data: request.toJson(),
      );
      return Organization.fromJson(response.data);
    } on DioException catch (e) {
      throw ApiException.fromResponse(
        e.response?.statusCode ?? 500,
        e.response?.data,
      );
    }
  }

  Future<void> deleteOrganization(String id) async {
    try {
      await _dio.delete(ApiConstants.organizationById(id));
    } on DioException catch (e) {
      throw ApiException.fromResponse(
        e.response?.statusCode ?? 500,
        e.response?.data,
      );
    }
  }
}

File: src/project-manager-web/src/features/organizations/repository/organization-repository.ts
import { apiClient } from "@/core/api/client";
import { ApiEndpoints } from "@/core/api/constants";
import { ApiException } from "@/core/api/exceptions";
import type { Organization, OrganizationDetail, CreateOrganizationRequest, UpdateOrganizationRequest } from "../models/types";
import { AxiosError } from "axios";

function handleError(error: unknown): never {
  if (error instanceof AxiosError) {
    throw ApiException.fromResponse(error.response?.status ?? 500, error.response?.data);
  }
  throw error;
}

export const organizationRepository = {
  async list(): Promise<Organization[]> {
    try {
      const response = await apiClient.get<Organization[]>(ApiEndpoints.organizations);
      return response.data;
    } catch (error) { handleError(error); }
  },

  async get(id: string): Promise<OrganizationDetail> {
    try {
      const response = await apiClient.get<OrganizationDetail>(ApiEndpoints.organizationById(id));
      return response.data;
    } catch (error) { handleError(error); }
  },

  async create(request: CreateOrganizationRequest): Promise<void> {
    try {
      await apiClient.post(ApiEndpoints.organizations, request);
    } catch (error) { handleError(error); }
  },

  async update(id: string, request: UpdateOrganizationRequest): Promise<void> {
    try {
      await apiClient.put(ApiEndpoints.organizationById(id), request);
    } catch (error) { handleError(error); }
  },

  async delete(id: string): Promise<void> {
    try {
      await apiClient.delete(ApiEndpoints.organizationById(id));
    } catch (error) { handleError(error); }
  },
};
